--- 
title: "Single-cell RNA-seq analysis with scrapper"
documentclass: book
bibliography: [ref.bib]
biblio-style: apalike
link-citations: yes
favicon: "favicon.ico"
description: "Or: how I learned to stop worrying and love the t-SNEs."
cover-image: "https://github.com/Bioconductor/BiocStickers/raw/devel/Bioconductor/Bioconductor-serial.gif"
github-repo: libscran/scrapbook
---

```{r, echo=FALSE}
library(BiocStyle)
.gh_url <- file.path('https://github.com', rmarkdown::metadata[['github-repo']])
knitr::opts_chunk$set(message=FALSE, warning=FALSE, error=FALSE)

curpath <- getwd()
while (curpath != (nextpath <- dirname(curpath))) {
    target <- file.path(curpath, "DESCRIPTION")
    if (file.exists(target)) {
        break
    }
    curpath <- nextpath
}
if (!file.exists(target)) {
    stop("failed to find a DESCRIPTION file")
}

info <- read.dcf(target)
authors <- eval(parse(text=info[,"Authors@R"]))
```

---
author: `r paste(paste('\n- name:', format(authors, include=c('given', 'family', 'role'))), collapse='')`
url: "`r .gh_url`"
---

# Introduction {-}

<a href="https://bioconductor.org"><img src="`r rmarkdown::metadata[["cover-image"]]`" width="200" alt="Bioconductor Sticker" align="right" style="margin: 0 1em 0 1em" /></a> 

Single-cell RNA-sequencing (scRNA-seq) - the name says it all, really.
Long story short, we isolate single cells and we sequence their transcriptomes to quantify the expression of each gene in each cell [@kolodziejczyk2015technology].
This allows us to explore heterogeneity in a cell population at the resolution of individual cells,
typically to identify subpopulations or states that would not be apparent from population-level (bulk) assays.
Since around 2014, scRNA-seq has emerged as one of the premier techniques for publishing papers in genomics and biomedicine.
Occasionally, it is even used to do some actual science.

This book describes a computational workflow for analyzing scRNA-seq data using the [R/Bioconductor](https://bioconductor.org) ecosystem.
Most of the heavy lifting is performed using the `r Biocpkg("scrapper")` package, while `r Biocpkg("scater")` handles the plotting.
We rely heavily on Bioconductor data structures like the `SingleCellExperiment` class,
so readers should check out the `r Biocpkg("SingleCellExperiment", "intro.html", label="associated documentation")` if they haven't already.
Each chapter is devoted to a particular step in the analysis where we provide its theoretical rationale, the associated code, and some typical results from real public datasets.
This includes:

- Quality control, to filter out cells that were damaged or not properly sequenced.
- Normalization, to remove cell-specific biases. 
- Feature selection, to identify genes with interesting biologial variation.
- Principal components analysis, to compact and denoise the data.
- Visualization, to generate the all-important Figure 1 of our manuscript.
- Clustering, to summarize the data into groups of similar cells. 
- Marker detection, to assign biological meaning to each cluster based on its upregulated genes.

Truth be told, you don't actually have to read this book if you don't care about how/why things are done.
If you just want to get those pretty pictures for your upcoming submission to _Nature_, just copy and paste the following into your R session:

```{r}
# Pulling out an example dataset.
library(scRNAseq)
sce.zeisel <- ZeiselBrainData()

# Running the full analysis pipeline.
library(scrapper)
is.mito.zeisel <- grep("^mt-", rownames(sce.zeisel))
res.zeisel <- analyze(counts(sce.zeisel), rna.subsets=list(MT=is.mito.zeisel))
sce.res.zeisel <- convertAnalyzeResults(res.zeisel)

# Visualizing the cluster assignments for each cell:
library(scater)
plotReducedDim(sce.res.zeisel, "tsne", colour_by="clusters")

# Looking at the top markers for cluster 1:
markers.1.zeisel <- reportGroupMarkerStatistics(res.zeisel$rna.markers, "1")
markers.1.zeisel <- markers.1.zeisel[order(markers.1.zeisel$cohens.d.mean, decreasing=TRUE),]
head(markers.1.zeisel[,c("mean", "detected", "cohens.d.mean")])
```

And that's it.
But if you want to look under the hood, it's worthwhile to keep reading as you might get some ideas for tweaking parameters to get better results.
Any questions can be posted at [Bioconductor support site](https://support.bioconductor.org) or the [GitHub page for this book](https://github.com/libscran/scrapbook).
